fix: Make File\Extension usable for non-existing files

edannenberg opened the pull request:
Currently the validator fails if the given filename is not readable. This prevents valid use cases like checking if a user configurable filename has an allowed file extension before creating said file.

As the validator only operates on the given string there is no need for the current readable check.

Also fixes File\ExcludeExtension which suffers from the same issue.

In theory it's a BC breaking change, personally I would consider any code that (ab)uses this validator as a file_is_readable validator to be smelly/buggy.

edannenberg committed:
fix: Make File\Extension usable for non-existing files

Currently the validator fails if the given filename is not readable.
This prevents valid use cases like checking if a user configurable
filename has an allowed file extension before creating said file.

As the validator only operates on the given string there is no
need for the current readable check.

Also fixes File\ExcludeExtension which suffers from the same issue.

weierophinney commented:
This is definitely a change in behavior; we have tests that verify that the validator returns `false` when the file does not exist (which you can see in the [test results](https://travis-ci.org/zendframework/zend-validator/jobs/525934909#L662)).

That said, looking through the code, you are absolutely correct - the validators are not testing any of the file _contents_, just the file _name_.

For BC purposes, this needs to be opt-in behavior, with the current behavior being the default. We can revisit that when we consider a new major version.

I'll push a change shortly to enable this.

weierophinney committed:
feat: adds allowNonExistentFile option to Extension and ExcludeExtension validators

This builds on the original patch. However, to avoid a BC break, this
patch adds a new option to each of the Extension and ExcludeExtension
file validators, `allowNonExistentFile`. By default, this flag is
`false`, preserving the existing behavior where a file must exist in
order to be validated. When `true`, the validator does not automatically
return `false` for non-existent files, and will continue to test its
extension.

weierophinney committed:
docs: adds documentation for new allowNonExistentFile option

weierophinney committed:
docs: adds CHANGELOG entry for #266

weierophinney commented:
I've pushed a change to make this optional behavior, and added documentation on how to enable the option.

weierophinney reviewed (approved):
With the change to make the new behavior optional, this can go into the next minor.
